crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs

```diff
@@ -66,57 +88,214 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
         let Some(builtin_type) = checker.semantic().resolve_builtin_symbol(expr) else {
             return;
         };
 
-        match builtin_type {
-            "int" => has_int = true,
-            "float" => has_float = true,
-            "complex" => has_complex = true,
-            _ => {}
-        }
+        numeric_flags.seen_builtin_type(builtin_type);
     };
 
+    // Traverse the union, and remember which numeric types are found.
     traverse_union(&mut find_numeric_type, checker.semantic(), annotation);
 
-    if has_complex {
-        if has_float {
-            checker.diagnostics.push(Diagnostic::new(
-                RedundantNumericUnion {
-                    redundancy: Redundancy::FloatComplex,
-                },
-                annotation.range(),
-            ));
+    let Some(redundancy) = Redundancy::from_numeric_flags(numeric_flags) else {
+        return;
+    };
+
+    // Traverse the union a second time to construct the fix.
+    let mut necessary_nodes: Vec<&Expr> = Vec::new();
+
+    let mut union_type = UnionKind::TypingUnion;
+    let mut remove_numeric_type = |expr: &'a Expr, parent: &'a Expr| {
+        let Some(builtin_type) = checker.semantic().resolve_builtin_symbol(expr) else {
+            // Keep type annotations that are not numeric.
+            necessary_nodes.push(expr);
+            return;
+        };
+
+        if matches!(parent, Expr::BinOp(_)) {
+            union_type = UnionKind::PEP604;
         }
 
-        if has_int {
-            checker.diagnostics.push(Diagnostic::new(
-                RedundantNumericUnion {
-                    redundancy: Redundancy::IntComplex,
-                },
-                annotation.range(),
-            ));
+        // `int` is always dropped, since `float` or `complex` must be present.
+        // `float` is only dropped if `complex`` is present.
+        if (builtin_type == "float" && !numeric_flags.contains(NumericFlags::COMPLEX))
+            || (builtin_type != "float" && builtin_type != "int")
+        {
+            necessary_nodes.push(expr);
+        }
+    };
+
+    // Traverse the union, and remember which numeric types are found.
+    traverse_union(&mut remove_numeric_type, checker.semantic(), annotation);
+
+    let mut diagnostic = Diagnostic::new(RedundantNumericUnion { redundancy }, annotation.range());
+    if checker.settings.preview.is_enabled() {
+        // Mark [`Fix`] as unsafe when comments are in range.
+        let applicability = if checker.comment_ranges().intersects(annotation.range()) {
+            Applicability::Unsafe
+        } else {
+            Applicability::Safe
+        };
+
+        // Generate the flattened fix once.
+        let fix = if let &[edit_expr] = necessary_nodes.as_slice() {
+            // Generate a [`Fix`] for a single type expression, e.g. `int`.
+            Fix::applicable_edit(
+                Edit::range_replacement(checker.generator().expr(edit_expr), annotation.range()),
```

Comment:
Is there a specific reason for using a generator over a text-based fix? Most fixes are text based to preserve existing formatting (and possibly comments).

Reply:
I've used the generator based on other rules, it could very well be that a text-based fix is more optimal.

Reply:
However, redundant expressions are still dropped. If they have the comment, it will be still lost.

Reply:
> I've used the generator based on other rules, it could very well be that a text-based fix is more optimal.

We generally prefer using text-based fixes if they are not too complicated because they preserve more comments and the formatting. But I don't think it's worth rewriting it now, or we can do so as a separate PR .